pkg/azuredx/azureauth/auth.go

```diff
@@ -53,55 +35,74 @@ func init() {
 }
 
 // ServiceCredentials provides authorization for cloud service usage.
-type ServiceCredentials struct {
+type ServiceCredentials interface {
+	ServicePrincipalAuthorization(ctx context.Context) (string, error)
+	QueryDataAuthorization(ctx context.Context, req *backend.QueryDataRequest) (string, error)
+}
+
+type ServiceCredentialsImpl struct {
 	models.DatasourceSettings
 	// HTTPDo is the http.Client Do method.
-	HTTPDo func(req *http.Request) (*http.Response, error)
-	// ServicePrincipalToken is the azidentity.ClientSecretCredential GetToken method.
-	ServicePrincipalToken func(ctx context.Context, opts azcore.TokenRequestOptions) (*azcore.AccessToken, error)
-	tokenCache            *cache
-	scope                 string
+	HTTPDo        func(req *http.Request) (*http.Response, error)
+	authority     azidentity.AuthorityHost
+	tokenProvider aztokenprovider.AzureTokenProvider
+	tokenCache    *cache
+	scopes        []string
```

Comment:
How come scopes is an array now? `getAzureScopes` doesn't return more than one scope.

(I see you had mentioned TODO #356 there, but didn't understand from the description of that ticket that more than one scope would be returned either.)

Reply:
While in this particular case ADX endpoint needs only a single scope (`.default` - is the default, catch-all scope), generally it's not always the case. An endpoint in theory _may_ require more than one scope at the same time, and access token can be issued for more than one scope.

Signature of the `GetAccessToken` func in `AzureTokenProvider` ([here](https://github.com/grafana/grafana-azure-sdk-go/blob/e079ca81c3fb6b309b37843345cf97f30aad4609/aztokenprovider/token_provider.go#L20)):

```go
type AzureTokenProvider interface {
	GetAccessToken(ctx context.Context, scopes []string) (string, error)
}
```

So I made `getAzureScopes` to return `[]string` array more for convention, as a type for representation of scopes, proper abstraction, rather than because it's technically required. The fact that `getAzureScopes` returns only one scope is implementation detail.